src/Frame.cc

```diff
@@ -1448,3 +1457,182 @@ void Frame::SaveFile(const std::string& root_folder, const CARTA::SaveFile& save
     save_file_ack.set_success(success);
     save_file_ack.set_message(message);
 }
+
+bool Frame::ExportCASAImage(casacore::ImageInterface<casacore::Float>& image, fs::path output_filename, casacore::String& message) {
+    bool success(false);
+
+    // Remove the old image file if it has a same file name
+    if (fs::exists(output_filename)) {
+        fs::remove_all(output_filename);
+    }
+
+    // Get a copy of all pixel data
+    casacore::IPosition start(image.shape().size(), 0);
+    casacore::IPosition count(image.shape());
+    casacore::Slicer slice(start, count);
+    casacore::Array<casacore::Float> temp_array;
+    image.doGetSlice(temp_array, slice);
+
+    // Construct a new CASA image
+    try {
+        auto out_image =
+            std::make_unique<casacore::PagedImage<casacore::Float>>(image.shape(), image.coordinates(), output_filename.string());
+        out_image->setMiscInfo(image.miscInfo());
+        out_image->setImageInfo(image.imageInfo());
+        out_image->appendLog(image.logger());
+        out_image->setUnits(image.units());
+        out_image->putSlice(temp_array, start);
+
+        // Create the mask for region
+        if (image.hasPixelMask()) {
+            casacore::Array<casacore::Bool> image_mask;
+            image.getMaskSlice(image_mask, slice);
+            out_image->makeMask("mask0", true, true);
+            casacore::Lattice<casacore::Bool>& out_image_mask = out_image->pixelMask();
+            out_image_mask.putSlice(image_mask, start);
+        }
+        success = true;
+    } catch (casacore::AipsError error) {
+        message = error.getMesg();
+    }
+
+    return success;
+}
+
+bool Frame::ExportFITSImage(casacore::ImageInterface<casacore::Float>& image, fs::path output_filename, casacore::String& message) {
+    // Remove the old image file if it has a same file name
+    return casacore::ImageFITSConverter::ImageToFITS(message, image, output_filename.string(), 64, casacore::False, casacore::True, -32,
+        1.0, -1.0, casacore::True, casacore::False, casacore::True, casacore::False, casacore::False, casacore::False, casacore::String(),
+        casacore::True);
+}
+
+void Frame::ValidateChannelStokes(std::vector<int>& channels, std::vector<int>& stokes, const CARTA::SaveFile& save_file_msg) {
+    auto image = GetImage();
+    auto image_shape = image->shape();
+
+    int channels_max = _spectral_axis > -1 ? image_shape[_spectral_axis] : 1;
+    int channels_start = 0;
+    int channels_stride = 1;
+    int channels_end = channels_max - 1;
+    int channels_length = channels_max;
+    if (save_file_msg.channels().size() > 0) {
```

Comment:
Please fix `repeated` typos in carta-protobuf messages.rst.txt

Reply:
Correct it.